[fastlane] add current_build_number action 🎯

hjanuschka opened the pull request:
as mentioned  in https://github.com/fastlane/fastlane/issues/6770 returns the current build number of either live_version or edit_version  - works only with spaceship-master (as this PR is required: https://github.com/fastlane/fastlane/pull/6785)

following ui:

``` ruby
lane :demo do
  # query the appstore version (ready for sale)
  build_number = current_build_number
  puts build_number

  #query the latest testflight upload
  build_number = current_build_number(live: false)
  puts build_number
end
```

edit: i think this is a pretty common feature, therefore my PR with a new action, instead of a plugin

hjanuschka committed:
add app_store_build_number action

KrauseFx commented:
```ruby
  #query the latest testflight upload
  build_number = current_build_number(live: false)
  puts build_number
```

Looking at this `current_build_number(live: false)` is the same as `latest_testflight_build_number`? Or is there any differences? 

Other than that I like the action and I feel like this is useful to many users 👍

KrauseFx commented on fastlane/docs/Actions.md:
```diff
@@ -6458,6 +6458,47 @@ Key | Description
 
+### current_build_number
+
+Fetches the current build number from either production or testflight train.
```
Should be `TestFlight`

hjanuschka replied:
done

KrauseFx commented on fastlane/lib/fastlane/actions/current_build_number.rb:
```diff
@@ -0,0 +1,102 @@
+      end
+
+      def self.authors
+        # So no one will ever forget your contribution to fastlane :) You are awesome btw!
```
You can remove this line

KrauseFx commented on fastlane/lib/fastlane/actions/current_build_number.rb:
```diff
@@ -0,0 +1,102 @@
+
+      def self.description
+        "Returns the current build_number of either live or edit version"
+      end
```
Can you also provide `self.details` with more details, in particular when to use this and when to use `latest_testflight_build_number`

hjanuschka replied:
done

KrauseFx commented on fastlane/spec/unused_options_spec.rb:
```diff
@@ -1,7 +1,7 @@
   describe Fastlane::Action do
     describe "No unused options" do
-      let (:all_exceptions) { %w(pilot appstore cert deliver gym match pem produce scan sigh snapshot supply testflight mailgun testfairy ipa import_from_git hockey deploygate crashlytics artifactory appledoc slather screengrab download_dsyms notification frameit set_changelog register_devices latest_testflight_build_number) }
+      let (:all_exceptions) { %w(pilot appstore cert deliver gym match pem produce scan sigh snapshot supply testflight mailgun testfairy ipa import_from_git hockey deploygate crashlytics artifactory appledoc slather screengrab download_dsyms notification frameit set_changelog register_devices latest_testflight_build_number current_build_number) }
```
Why does this action have an unused action?

hjanuschka replied:
the `:team_id` is only used in the verify_block, wich is not detected by the spec.

```
verify_block: proc do |value|
                                         ENV["FASTLANE_ITC_TEAM_NAME"] = value.to_s
                                       end

```

hjanuschka commented:
yeah its pretty much the same return value as in `latest_testflight_build_number` -first i tried to modify the `latest_testflight_build_number` - to return the build_number from live_version.
but it would be too much spagetti. so i decided to do a new action, and found that it maybe usefull to query both variations, as the spaceship extension is on `AppVersion` 👍

KrauseFx commented:
How about we call the new action `current_build_number` from the TestFlight action so that we have only one code to maintain?

hjanuschka commented:
thats a great idea, i am working on it!

hjanuschka commented:
@KrauseFx  - ok  i am  a bit unsure of the refactor, as the testflight variation uses the build_trains - and i am a bit scared that it will break too much.

hjanuschka commented:
@KrauseFx  ok after a few hours of sleeping, i think i got a solution, will refactor both actions tonight - to share as much code as possible - without breaking BC

hjanuschka commented:
@KrauseFx  ok refactored the `latest_testflight_build_number` to be basically like an alias (with a forced `live: false`, and a bc sharedValue) - i think it should'nt break anything.


```
lane :demo do
  b = current_build_number(app_identifier: "krone.at-fivexfive", username: "helmut@januschka.com")
  puts b

  b = current_build_number(live: false, app_identifier: "krone.at-fivexfive", username: "helmut@januschka.com")
  puts b

  latest_testflight_build_number(app_identifier: "krone.at-fivexfive", username: "helmut@januschka.com")
  
  b = current_build_number(live: false, app_identifier: "krone.at-fivexfive", username: "helmut@januschka.com", version: "1.2.9")
  puts b
  
  latest_testflight_build_number(app_identifier: "krone.at-fivexfive", username: "helmut@januschka.com", version: "1.2.9")

end

```

tested with the lane above, and it did work :)

milch commented on fastlane/lib/fastlane/actions/current_build_number.rb:
```diff
@@ -0,0 +1,142 @@
+      end
+
+      def self.details
+        "Returns the current build number of either the live or testflight version - it is usefull for getting the build_number of the current, ready-for-sale, app version, it also works on non-live testflight version. If you need to handle more build-trains please see `latest_testflight_build_number`"
```
`usefull` -> `useful`

milch replied:
`current, ready-for-sale, app version, it also works` -> `current or ready-for-sale app version, and it also works`

milch commented on fastlane/lib/fastlane/actions/current_build_number.rb:
```diff
@@ -0,0 +1,142 @@
+              version_number = testflight_version
+            else
+              UI.message("You have to specify a new version number: ")
+              version_number = STDIN.gets.strip
```
Use UI.input [as described here](https://github.com/fastlane/fastlane/blob/master/fastlane/docs/UI.md)

milch commented on fastlane/lib/fastlane/actions/current_build_number.rb:
```diff
@@ -0,0 +1,142 @@
+      end
+
+      def self.is_supported?(platform)
+        true
```
`is_supported?` is supposed to return true for the action target's platform - so in this case `[:ios, :mac].include?(platform)` is probably more appropriate

milch reviewed (changes requested):
In general, I don't like the name `current_build_number`. Just from that name I'd guess that it's only fetching the current build number from the Xcode project or whatever.

fotiDim commented:
Maybe in order to be consistent with `latest_testflight_build_number` it should be named `latest_app_store_build_number` or `latest_live_build_number`. 

Just a small question though... What if there is a build in `pending developer release` state? Would this build number be returned or the public one?

hjanuschka commented:
* @milch feedback done! ( i love the gone `Actions.md` 🎉 )
  * rebased current master and squased all commits

about the name how about `app_store_build_number` - as it is not always the latest! (you can set `:version`  for a specific train)

@KrauseFx  can we get a core-team decision about the name? - so i can refactor it?


@fotiDim if your release is pending dev release - it will return the build number from published one, as of the `state` - the "pending dev release` is not live

milch commented:
@hjanuschka Thanks! 

I like `app_store_build_number`, it makes it pretty clear that this is not something that is read from a local file or something.

KrauseFx commented on fastlane/lib/fastlane/actions/current_build_number.rb:
```diff
@@ -0,0 +1,141 @@
+    class CurrentBuildNumberAction < Action
+      def self.run(params)
+        require 'spaceship'
+        credentials = CredentialsManager::AccountManager.new(user: params[:username])
```
This is not necessary, check out the other actions like `DownloadDsymsAction`. I just checked, and this code has been here before, but let's get rid of it and do like `DownloadDsymsAction`

KrauseFx commented on fastlane/lib/fastlane/actions/latest_testflight_build_number.rb:
```diff
@@ -69,6 +33,12 @@ def self.available_options
         user ||= CredentialsManager::AppfileConfig.try_fetch_value(:apple_id)
 
         [
+          FastlaneCore::ConfigItem.new(key: :live,
```
This value gets overwritten in any case, so that's not good

hjanuschka replied:
done  - with default value `false` and not overwriting it, basically the `latest_testflight_build_number` - should be deprecated - but should work as a BC shim!

KrauseFx commented on fastlane/lib/fastlane/actions/latest_testflight_build_number.rb:
```diff
@@ -69,6 +33,12 @@ def self.available_options
+                                       env_name: "CURRENT_BUILD_NUMBER_LIVE",
+                                       description: "Query the live version (ready-for-sale)",
+                                       optional: true,
+                                       default_value: true),
```
We overwrite the default value?

KrauseFx replied:
Let's do it like this: Make the default value `false`, and not set it on line 11 (just remove line 11)

KrauseFx commented:
Thanks everyone, this looks great already, I added some more comments before this is ready 👍

hjanuschka commented:
what a PR! 😵  - thx for all your help! 🎉 

  * rebased current master, squashed commits
  * @KrauseFx  done the feedback
  * renamed action to `app_store_build_number`


test lanes:

```ruby
lane :demo do
  b = app_store_build_number(app_identifier: "krone.at-fivexfive", username: "helmut@januschka.com")
  puts b

  b = app_store_build_number(live: false, app_identifier: "krone.at-fivexfive", username: "helmut@januschka.com")
  puts b

  latest_testflight_build_number(app_identifier: "krone.at-fivexfive", username: "helmut@januschka.com")
  
  b = app_store_build_number(live: false, app_identifier: "krone.at-fivexfive", username: "helmut@januschka.com", version: "1.2.9")
  puts b
  
  latest_testflight_build_number(app_identifier: "krone.at-fivexfive", username: "helmut@januschka.com", version: "1.2.9")

end

```

KrauseFx commented on fastlane/lib/fastlane/actions/app_store_build_number.rb:
```diff
@@ -0,0 +1,141 @@
+
+      def self.example_code
+        [
+          'appstore_build_number'
```
There should be a second example with more complex parameters, similar to what we do in most other actions: https://github.com/fastlane/fastlane/blob/master/fastlane/lib/fastlane/actions/sigh.rb#L71-L80

KrauseFx commented on fastlane/lib/fastlane/actions/app_store_build_number.rb:
```diff
@@ -0,0 +1,141 @@
+        UI.message("Login successful")
+
+        app = Spaceship::Tunes::Application.find(params[:app_identifier])
+        build_nr = nil
```
I don't think we need to assign this to `nil`, right?

hjanuschka commented:
@KrauseFx issues have been addressed.

milch commented:
Thanks 👍

fotiDim commented:
Great work guys! When do we expect a new fastlane release that includes those changes?

hjanuschka commented:
@fotiDim  1.107.0 has it -> https://github.com/fastlane/fastlane/releases/tag/1.107.0 🎉
